[GSoC] Cloudplugin

shiva-kannan opened the pull request:
**Link the Issue(s) this Pull Request is related to.**
No associated issue for now. Will link it back when created

**Summarize your change.**
- First pass at a fully functional cloud plugin that can be launched from CueCommander 
- Only connects to Google Cloud for now
- Authentication, for now, is using Application Default Creds setup on GCP (hardcoded into gce_api.py)


- Can do the following using the plugin 
> See all the available cloud groups setup on GCE
> Create a cloud group with the widget (uses the templates/vms already existing with your cloud project)
> Resize the group from within the widget 
> Delete a particular cloud group

bcipriano commented on pycue/tests/cloudinstancegroup_test.py:
```diff
@@ -0,0 +1,78 @@
+#!/usr/bin/env python
```
Regarding file naming/location: the convention we follow is that test classes and their directory structure should mirror the class under test.

In this case, the class under test is `opencue.cloud.gce_api.GoogleCloudGroup` so this file should be `pycue/tests/cloud/gce_api_test.py`.

You've named your test class `GoogleCloudGroupTest` which already follows the convention for test class names. 👍

bcipriano commented on pycue/opencue/cloud/gce_api.py:
```diff
@@ -0,0 +1,145 @@
+credentials = GoogleCredentials.get_application_default()
+service = discovery.build('compute', 'v1', credentials=credentials)
+project = 'gsoc-opencue-test-bed'
+zone = 'us-central1-a'
```
Let's avoid having these values as globals. The main problem here is that by importing this file/class, you automatically trigger some code logic to run (fetching credentials and building the service) which can lead to some unexpected behavior and make the class difficult to test.

Normally the pattern would be to have them as arguments to the constructor. Then in the "real" code, pass in the values you have here. Then in the test code, you could pass in whatever values you wanted -- mocks/dummy values/etc.

That won't really work here, because we're implementing an abstract class and we want the constructor params to be the same for different cloud providers. For example the AWS version of this class probably won't use the same project/zone setup, so we don't want to use those parameters.

How about something like this -- move these values to be class-level variables in the `GoogleCloudGroup` class. Add a `connect` method to the class which defines their values. That way the GUI code can instantiate the class and call `connect()`, and the test code can skip calling `connect()`, mocking the values instead.

shiva-kannan replied:
I revisited the way I have setup the CloudGroup class. Since connect() is more of a CloudManager job I have re-structured the methods a bit and pulled out these globals into attributes of the implemented provider Manager class. 
I will discuss this restructuring a bit tomorrow during the check-in but like you mentioned it helped me avoid any google related code completely while setting up tests

bcipriano commented on pycue/opencue/cloud/gce_api.py:
```diff
@@ -0,0 +1,145 @@
+from .api import CloudInstanceGroup
+
+from googleapiclient import discovery
+from oauth2client.client import GoogleCredentials
```
Go ahead and add these libraries to the top-level `requirements.txt` so we get tests running in the CI platform. It's not ideal to install Google-specific libraries if folks aren't using the Google versions of these plugins, but we don't have a good way to supplement optional libraries at the moment -- we can revisit that later.

shiva-kannan replied:
I had to added both `google-python-client` and `oauth2client` and seems like the new test  I setup ran successfully!

bcipriano commented on pycue/tests/cloudinstancegroup_test.py:
```diff
@@ -0,0 +1,78 @@
+        self.http = HttpMock(os.path.join(GOOGLE_RESOURCES_DIR, "compute-discovery.json"),
+                             {'status': '200'})
+        api_key = "test_api_key"
+        self.service = discovery.build('compute', 'v1', http=self.http, developerKey=api_key)
```
I think that we're including too much third-party Google logic in these tests.

Ideally we don't want to be calling into Google-owned modules at all. We're not testing the cloud provider modules themselves; we only want to make sure our own code is behaving as expected. So we really want to being thinking more along the lines of "if the Google API were to return X as its output, would my code handle that result properly?" So rather than calling `discovery.build` and then mocking out pieces of it, mock the whole class.

This is related to the comment above about global variables. Once the `service` variable is no longer a global, you can use `mock.patch.object` to set up `GoogleCloudGroup.service` as a `Mock` (in lieu of calling `connect()` as the main code does/will do). Then we can mock out `service` as needed without ever calling into those Google modules.

Hopefully that makes sense, can explain more during our next meeting.

shiva-kannan replied:
I have redone the setup for tests for this module. I'll go through these with you during our check-in and then finish up the other tests if this structure looks good

shiva-kannan commented on cuegui/cuegui/CloudGroupDialog.py:
```diff
@@ -0,0 +1,116 @@
+#  Copyright (c) 2018 Sony Pictures Imageworks Inc.
```
This should be fixed with the new commits

shiva-kannan commented on cuegui/cuegui/CloudGroupDialog.py:
```diff
@@ -0,0 +1,116 @@
+from PySide2 import QtGui
+from PySide2 import QtWidgets
+
+from opencue.cloud.api import CloudManager
```
Fixed the imports for all my modules as well. Let me know if the changes look good

bcipriano commented on pycue/opencue/cloud/gce_api.py:
```diff
@@ -0,0 +1,161 @@
+
+    @opencue.cloud.gce_exception_util.googleRequestExceptionParser
+    def create_managed_group(self, name, size, template):
+        # TODO : Use request ID to handle multiple create button clicks
```
Elsewhere in CueGUI this is handled at the plugin level -- a creation request gets added to some internal cache within the widget, then the widget's tick method iterates over the requests and takes action on each of them. In that logic you could do things like de-duping requests as well.

Doesn't need to happen in this PR, just wanted to note for later.

bcipriano commented on pycue/tests/cloud/gce_api_test.py:
```diff
@@ -0,0 +1,137 @@
+        # __getitem__.side_effect is used to make the return value of mock
+        # list call subscriptable since the actual code uses it like a dictionary
+        test_response = {
+            "items": [
```
Make sure to add a second item to this response.

bcipriano commented on pycue/tests/cloud/gce_api_test.py:
```diff
@@ -0,0 +1,137 @@
+                }
+            ]
+        }
+        connect_mock.return_value = None
```
Do you need this `return_value`? If you just want to prevent connect from being called, mocking it should be sufficient unless the calling code requires a specific response.

If you just need to prevent something from being called and don't need to set up a return value, you can set it up like:

```
@mock.patch.object(opencue.cloud.gce_api.GoogleCloudManager, "connect", autospec=True, new=mock.MagicMock(return_value=False))
```

(with or without `return_value`, depending on what's needed).

This will prevent the mock from being passed in as an input to the test method, which can keep the code a bit cleaner.

shiva-kannan replied:
Yes, that makes sense. After moving `connect` out of the constructor I wouldn't even need to patch it like you mentioned. I forgot to move it out of the tests. Thanks for catching that!

bcipriano commented on requirements.txt:
```diff
@@ -12,3 +12,5 @@ pyfakefs==3.6
 simplejson==3.16.0
 six==1.11.0
+google-api-python-client==1.9.1
+oauth2client==4.1.3
```
Please alphabetize this file as well.

shiva-kannan commented:
> Took a first pass through. This is looking very good, most of these comments are quite minor -- thanks!

@bcipriano addressed all of these comments with the latest commit. Let me know if it works

bcipriano commented on cuegui/cuegui/CloudGroupDialog.py:
```diff
@@ -0,0 +1,124 @@
+        QtWidgets.QComboBox.__init__(self, parent)
+        self._templates = {}
+
+    def refresh(self, cloud_group):
```
Sorry, to clarify -- you want `:type` and `:param` lines for each parameter.

For this method that might look like:

```
:type cloud_group: opencue.cloud.api.CloudInstanceGroup
:param cloud_group: cloud group to be refreshed
```

You can skip `:return:` when there's nothing returned.

bcipriano commented on pycue/opencue/cloud/api.py:
```diff
@@ -0,0 +1,144 @@
+    @staticmethod
+    def get_registered_providers():
+        """
+        :return:
```
Make sure to go through and document all the methods here. What's returned here is particularly important.

bcipriano commented on pycue/opencue/cloud/api.py:
```diff
@@ -0,0 +1,134 @@
+    def create_managed_group(self, name, size, template):
+        """
+        Creates a cloud group with the given template/image from cloud
+        :return:
```
Looks good here, but make sure to hit all the other methods, split into `:type`/`:param`, and document the return value.

shiva-kannan commented:
> All comments resolved, I think we're good to merge here!
> 
> Please switch the PR out of draft mode when you have a chance.

Done! Let me know if I need to do anything else
